torchvision/models/convnext.py

```diff
@@ -0,0 +1,201 @@
+from functools import partial
+from typing import Any, Callable, Dict, List, Optional, Sequence
+
+import torch
+from torch import nn, Tensor
+from torch.nn import functional as F
+
+from .._internally_replaced_utils import load_state_dict_from_url
+from ..ops.misc import ConvNormActivation
+from ..ops.stochastic_depth import StochasticDepth
+from ..utils import _log_api_usage_once
+
+
+__all__ = [
+    "ConvNeXt",
+    "convnext_tiny",
+]
+
+
+model_urls: Dict[str, Optional[str]] = {}
+
+
+class LayerNorm2d(nn.LayerNorm):
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        self.channels_last = kwargs.pop("channels_last", False)
+        super().__init__(*args, **kwargs)
+
+    def forward(self, x):
+        if not self.channels_last:
+            x = x.permute(0, 2, 3, 1)
```

Comment:
At TorchVision, we try to reuse as much as possible standard PyTorch components. I understand that in your original implementation, you provide a custom implementation for [channels first](https://github.com/facebookresearch/ConvNeXt/blob/main/models/convnext.py#L138-L143). Could you talk about the performance degradation you've experienced and how much this was that lead you to reimplementing it?

Reply:
The thing is we don't have a standard PyTorch component for channels_first LN. Here are some discussions with Ross Wightman (https://twitter.com/wightmanr/status/1481383509142818817?s=20). In Ross's timm implementation, he has a presumably better LN implementation for channel_first. See here: https://github.com/rwightman/pytorch-image-models/blob/b669f4a5881d17fe3875656ec40138c1ef50c4c9/timm/models/convnext.py#L109

Reply:
Thanks for the references. I wasn't aware of the concurrent discussions. I'll have a look to measure on our side. Ideally I would like to reuse the existing kernels as much as possible, unless there is a big gap in performance to justify a custom implementation.

Reply:
Hi @datumbox,
In this post I try to explain the motivation for our specific choices, in case it's still helpful: https://github.com/facebookresearch/ConvNeXt/issues/18#issuecomment-1014390947 
Thanks for your work on incorporating ConvNeXt!

Reply:
@liuzhuang13 Thanks for the reply. Is it fair to say that the approach that we follow here is expected to be 0-5% slower than the optimum? I haven't had the chance to run benchmarks but that's what I understand from your note.

FYI, the issue is that TorchVision has a common API across all models to accept the `norm_layer` as a parameter which is going to be tricky to support if I switch the 1x1 convs to linear. Where or not we will do this, depends on the speed impact.

Reply:
Yes, I just tried switching to conv layers in residual blocks, and using permutation + LN in all cases. I found combined together they cause 20-30% slowdown in inference @ 224 resolution for ConvNeXt-T/S. However, in ConvNeXt-B at 224 or any model at 384 resolution, it seems as fast as our original impl. This is on V100s, I cannot say much on other platforms though. It is indeed a bit strange to me.

Reply:
Thanks a lot for looking into it @liuzhuang13. 20-30% slow sounds very large and we would probably want to make it faster. I'll run benchmarks later and share with you any findings.